File: 92. Reverse Linked List II/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        # attach dummy head
        node = ListNode(None)
        node.next = head
        head = node
        
        # reach mth node
        prev = head
        node = head.next
        while m != 1:
            prev = node
            node = node.next
            m -= 1
            n -= 1
        
        # reverse the nodes
        tail1 = prev
        tail2 = node
        prev = None
        while n:
            tmp = node.next
            node.next = prev
            prev = node
            node = tmp
            n -= 1
            
        # attach all parts together
        tail1.next = prev
        tail2.next = node
        return head.next
        

File: 92. Reverse Linked List II/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        // insert dummy node
        ListNode node = new ListNode(0);
        node.next = head;
        head = node;
        
        // arrive at the mth node
        ListNode prev = head;
        node = head.next;
        
        while (m > 1) {
            m--;
            n--;
            prev = node;
            node = node.next;
        }
        
        // reverse m-n nodes
        ListNode tail = prev;
        ListNode head2 = node;
        ListNode tmp;
        prev = null;
        while (n > 0) {
            n--;
            tmp = node.next;
            node.next = prev;
            prev = node;
            node = tmp;
        }
        
        // reattach and return
        tail.next = prev;
        head2.next = node;
        return head.next;
    }
}
